// python3 -m grpc_tools.protoc -I . --python_out=. --grpc_python_out=. movie.proto

syntax = "proto3";

message Movie {
    string title = 1;
    double rating = 2;
    optional Review review = 3;
    int32 release_year = 4;
    optional string review_date = 5;
    optional bool redux = 6;
    int32 id = 7;
    string imdb_id = 8;

    message Review {
        
        message Person {
            string name = 1;
            optional string imdb_id = 2;
            optional string comments = 3; // Deprecated
        }

        message GenericCategory {
            optional string rating = 1;
            optional string comments = 2;
        }

        message Direction {
            repeated Person director = 1;
            optional string rating = 2;
            optional string comments = 3;
        }
        
        message Acting {
            message Performance {
                Person actor = 1;
                string rating = 2;
                string comments = 3;
            }
            repeated Performance performance = 1;
            optional GenericCategory cast = 2;
            optional string rating = 3;
        }
        
        message Story {
            repeated Person writer = 1;
            optional string rating = 2;
            optional string comments = 3;
        }

        message Screenplay {
            repeated Person writer = 1;
            optional string rating = 2;
            optional string comments = 3;
        }

        message Score {
            repeated Person composer = 1;
            optional string rating = 2;
            optional string comments = 3;
        }

        message Cinematography {
            repeated Person cinematographer = 1;
            optional string rating = 2;
            optional string comments = 3;
        }

        message Editing {
            repeated Person editor = 1;
            optional string rating = 2;
            optional string comments = 3;
        }

        optional Direction direction = 1;
        optional Acting acting = 2;
        optional Story story = 3;
        optional Screenplay screenplay = 4;
        optional Score score = 5;
        optional Cinematography cinematography = 6;
        optional Editing editing = 7;
        optional GenericCategory sound = 8;
        oneof visuals {
            GenericCategory visual_effects = 9;
            GenericCategory animation = 10;
        }
        optional GenericCategory production_design = 11;
        optional GenericCategory makeup = 12;
        optional GenericCategory costumes = 13;
        optional string plot_structure = 14;
        optional string pacing = 15;
        optional string climax = 16;
        optional string tone = 17;
        optional string final_notes = 18;
        optional string overall = 19;
    }
}

// Free Format
message MovieFree {
    string title = 1;
    double rating = 2;
    optional string review = 3;
    int32 release_year = 4;
    optional string review_date = 5;
    optional bool redux = 6;
    int32 id = 7;
    string imdb_id = 8;
}

message MovieOldFormat {
    string title = 1;
    double rating = 2;
    optional Review review = 3;
    int32 release_year = 4;
    optional string review_date = 5;
    optional bool redux = 6;
    int32 id = 7;
    string imdb_id = 8;

    message Review {
        
        message Person {
            string name = 1;
            optional string imdb_id = 2;
            optional string comments = 3;
        }

        message Direction {
            repeated Movie.Review.Person director = 1;
            optional string comments = 2;
        }
        
        message Acting {
            repeated Person actor = 1;
            optional string cast = 2;
            optional string comments = 3;
        }
        
        message Story {
            repeated Movie.Review.Person writer = 1;
            optional string comments = 2;
        }

        message Screenplay {
            repeated Movie.Review.Person writer = 1;
            optional string comments = 2;
        }

        message Score {
            repeated Movie.Review.Person composer = 1;
            optional string comments = 2;
        }

        message Cinematography {
            repeated Movie.Review.Person cinematographer = 1;
            optional string comments = 2;
        }

        message Editing {
            repeated Movie.Review.Person editor = 1;
            optional string comments = 2;
        }

        optional Direction direction = 1;
        optional Acting acting = 2;
        optional Story story = 3;
        optional Screenplay screenplay = 4;
        optional Score score = 5;
        optional Cinematography cinematography = 6;
        optional string sound = 7;
        optional Editing editing = 8;
        optional string visual_effects = 9;
        optional string production_design = 10;
        optional string makeup = 11;
        optional string costumes = 12;
        optional string plot_structure = 13;
        optional string pacing = 14;
        optional string climax = 15;
        optional string tone = 16;
        optional string final_notes = 17;
        optional string overall = 18;
    }
}